### ISSUE TITLE
Handling output in souffle converter

### ISSUE THREAD

**Issue by remysucre** [ryzhyk/differential-datalog#112]
It appears [`convert.py`](https://github.com/ryzhyk/differential-datalog/blob/master/test/souffle/convert.py) ignores the `output` annotations in souffle declarations, e.g. [this line](https://github.com/ryzhyk/differential-datalog/blob/95e358dfc6600f5a60eb35386132becef22857c6/test/souffle/self-contained.dl#L471) in `self-contained.dl`. 

I attempted to change `convert.py` to add `dump` statements in the converted `.dat` file, but got confused how the global [`relations`](https://github.com/ryzhyk/differential-datalog/blob/95e358dfc6600f5a60eb35386132becef22857c6/test/souffle/convert.py#L12) works. Is it strictly used for input relations? I see it's used in [`relation_name`](https://github.com/ryzhyk/differential-datalog/blob/95e358dfc6600f5a60eb35386132becef22857c6/test/souffle/convert.py#L156), which is used in methods that don't seem to be specific to inputs.

**Comment by ryzhyk** [comment]
Glad to hear you're trying out this feature.

We could ask @mbudiu-vmw to add commands to dump output relations in the end of the `.dat` file; however a better solution is coming in the `ovn-integration` branch that I should really merge soon.

Until few days ago, DDlog did not have the notion of output relations, or rather all relations were outputs. This is why the script ignores output annotations. However, it turns out that handling actual outputs differently from other (internal) relations has some performance benefits. So we introduced an `output` qualifier in relation declaration, which is symmetric to the `input` qualifier. We also modified the `convert.py` script to carry this qualifier over from Souffle to DDlog. One effect of the `output` qualifier is that the `dump;` command in `.dat` file without arguments only prints output relations.

So in summary if you switch to the ovn-integration branch and add a single `dump;` command at the end of the file, you should get a dump of output tables only. Does this help? Shall I modify `convert.py` to add the command?

Another issue that you will likely run into next is that the format in which DDlog dumps tables is neither very human-friendly nor very easy to pass to other tools. It is not hard to produce a specialized output adapter that dumps records in whatever format you prefer. This is not documented yet, and the relevant APIs are still in flux, but I should be able to help. We can talk offline...

**Comment by remysucre** [comment]
> So in summary if you switch to the ovn-integration branch and add a single dump; command at the end of the file, you should get a dump of output tables only. Does this help? Shall I modify convert.py to add the command?

Yes, and sure! 

> We can talk offline... (RE output API)

Sounds good. 

Another thing is that the `.decl ... output`syntax is deprecated in souffle, and `.output rel` should be used instead. See [doc](https://souffle-lang.github.io/docs/io/). The current code works for now, but I'm happy to help tweak that.

**Comment by ryzhyk** [comment]
I pushed the change to the ovn-integration branch. The patch also removes the `--no-store` flag, so running
```
stack test --ta '-p souffle'
```
now produces `test/souffle/souffle.dump` with output relations in a semi-readable format.

Fixing the script would definitely be helpful! @mbudiu-vmw wrote it very quickly just so that we could try DDlog on the one souffle example we obtained. Making it more robust by supporting the new syntax, removing hard-wired file names, etc. would be nice. Perhaps Mihai can point out other issues he is aware of. Mihai, we now support antijoins with wildcards. Does the script need any adjustments for that?

**Comment by ryzhyk** [comment]
@remysucre , which language would you like to use DDlog from in your use cases?

**Comment by remysucre** [comment]
Like either Haskell or Java, or through Doop.

**Comment by remysucre** [comment]
That would be helpful; wouldn't a Haskell API require little effort, or does that already exist?

**Comment by ryzhyk** [comment]
There's no Haskell API at the moment, but it should be possible to wrap the C API in Haskell. Would you be interested in doing this (once the C API is ready)?  The C API is going to be very simple, only using primitive types like `int` and `char*`, so the Haskell wrapper should be straightforward. Alternatively, I can put it on my todo list, but it may take me some time to get around to it...

Just to clarify, although DDlog itself is written in Haskell, the generated code is Rust, and so the only "native" interface to a DDlog program is the Rust API.

**Comment by remysucre** [comment]
Ah I see. I can justify the time to write the Haskell API if I end up using Haskell; right now I'm writing Java, but in a month or so I might switch to Haskell.

**Comment by ryzhyk** [comment]
Makes sense.  I am closing the issue. We can discuss the Haskell API when it becomes relevant.

**Comment by mbudiu-vmw** [comment]
Today I have two paper deadlines and I will be traveling the rest of the week, so I won't be able to work on the Java API this week.

**Comment by luc-tielen** [comment]
I know this comment is very late, but [souffle-haskell](https://github.com/luc-tielen/souffle-haskell) provides a Haskell wrapper around the C API. 😃
